dubbo-remoting/dubbo-remoting-zookeeper/src/main/java/org/apache/dubbo/remoting/zookeeper/support/AbstractZookeeperTransporter.java

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+package org.apache.dubbo.remoting.zookeeper.support;
+
+import org.apache.dubbo.common.Constants;
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.common.logger.Logger;
+import org.apache.dubbo.common.logger.LoggerFactory;
+import org.apache.dubbo.remoting.zookeeper.ZookeeperClient;
+import org.apache.dubbo.remoting.zookeeper.ZookeeperTransporter;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * AbstractZookeeperTransporter is abstract implements of ZookeeperTransporter.
+ * <p>
+ * If you want to extends this, implements createZookeeperClient.
+ */
+public abstract class AbstractZookeeperTransporter implements ZookeeperTransporter {
+    private static final Logger logger = LoggerFactory.getLogger(ZookeeperTransporter.class);
+    private final Map<String, ZookeeperClient> zookeeperClientMap = new ConcurrentHashMap<>();
+
+    /**
+     * share connnect for registry, metadata, etc..
+     *
+     * @param url
+     * @return
+     */
+    public ZookeeperClient connect(URL url) {
+        ZookeeperClient clientData;
+        List<String> addressList = getURLBackupAddress(url);
+        // The field define the zookeeper server , including protocol, host, port, username, password
+        if ((clientData = fetchAndUpdateZookeeperClientCache(url, addressList)) != null) {
+            logger.info("Get result from map for the first time when invoking zookeeperTransporter.connnect");
+            return clientData;
+        }
+        ZookeeperClient zookeeperClient = null;
+        // avoid creating too many connections， so add lock
+        synchronized (zookeeperClientMap) {
+            if ((clientData = fetchAndUpdateZookeeperClientCache(url, addressList)) != null) {
+                logger.info("Get result from map for the second time when invoking zookeeperTransporter.connnect");
+                return clientData;
+            }
+
+            zookeeperClient = createZookeeperClient(createServerURL(url));
+            logger.info("Get result by creating new connection when invoking zookeeperTransporter.connnect. ");
+            writeToClientMap(addressList, zookeeperClient);
+        }
+        return zookeeperClient;
+    }
+
+    /**
+     * @param url the url that will create zookeeper connection .
+     *            The url in AbstractZookeeperTransporter#connect parameter is rewritten by this one.
+     *            such as: zookeeper://127.0.0.1:2181/org.apache.dubbo.remoting.zookeeper.ZookeeperTransporter
+     * @return
+     */
+    protected abstract ZookeeperClient createZookeeperClient(URL url);
+
+    ZookeeperClient fetchAndUpdateZookeeperClientCache(URL url, List<String> addressList) {
+
+        ZookeeperClient zookeeperClientData = null;
+        for (String address : addressList) {
```

Comment:
This can written as 
```
priavte ZookeeperClient fetchAndUpdateZookeeperClientCache(URL url, List<String> addressList) {
Optional<ZookeeperClient> record=addressList.stream().filter(parameters::containsKey).findFirst();
if(record.isPresent()) {
    writeToClientMap(addressList, record.get());
}
return null;
}
```

Reply:
This is Stream usage, donot modify this time.